### PR TITLE
feat: viewsets and apiviews tests

### LINKED ISSUE #22: Pytest tests for crud endpoints
pick one viewset and one model with generic api views and write tests. Try using factories to create test instances

### PR DESCRIPTION
**Pull request by adlbrtbsr**
added tests for crud endpoints

### CODE REVIEW

File: core/tests/test_views.py [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+        )
+
+        assert response.status_code == 400
+        assert response.json() == {
+            "name": ["Ensure this field has no more than 50 characters."]
+        }
+
+    def test_genre_get_list(self, client):
+        genre1 = GenreFactory.create()
+        genre2 = GenreFactory.create()
```
Comment (maclewan) [inline core/tests/test_views.py]: any `.create()` calls are redundant, `genre1 = GenreFactory()` is enough

File: core/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+
+    def test_genre_post_invalid_name(self, client):
+        response = client.post(
+            self.url_list, data={"name": "a" * 51}, content_type="application/json"
+        )
+
+        assert response.status_code == 400
+        assert response.json() == {
+            "name": ["Ensure this field has no more than 50 characters."]
+        }
```
Comment (maclewan) [inline core/tests/test_views.py]: Imo there is no point in testing serializer/model validation. We need to trust default django/drf behavior, because otherwise we would need to test each field for each possible invalid scenario

File: core/tests/test_views.py [hunk 3]
```diff
@@ -0,0 +1,659 @@
...
+        assert response_data["name"] == serialized["name"]
+        assert response_data["is_for_adults"] == serialized["is_for_adults"]
+        assert Genre.objects.filter(name=serialized["name"]).exists()
+        assert set(response_data.keys()) == {"id", "name", "is_for_adults"}
+
+    def test_genre_post_blank(self, client):
+        response = client.post(self.url_list, data={}, content_type="application/json")
+
+        assert response.status_code == 400
+        assert response.json() == {"name": ["This field is required."]}
```
Comment (maclewan) [inline core/tests/test_views.py]: Kinda probably same as below, although with empty fields, tests like that might be useful, in case we want to ensure that some critical fields are always required by serializer/model. Also reverse scenario test makes even more sense if field might be empty for some reason. I would leave this test tho, but be aware, if someone ever asks why writing such kind of test

File: core/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+
+        response = client.post(
+            self.url_list, data=serialized, content_type="application/json"
+        )
+        response_data = response.json()
+
+        assert response.status_code == 201
+        assert response_data["name"] == serialized["name"]
+        assert response_data["is_for_adults"] == serialized["is_for_adults"]
+        assert Genre.objects.filter(name=serialized["name"]).exists()
```
Comment (maclewan) [inline core/tests/test_views.py]: not sure if we have id in the model, if so, I would always default to checking by id, even if other fields are unique. Now they are, in the future they might not be, id/pk always id unique. tl;dr idiot proofing tests

File: core/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,659 @@
...
+        assert set(response_data.keys()) == {"id", "name", "is_for_adults"}
```
Comment (maclewan) [inline core/tests/test_views.py]: I would just create separate test to cover model serialization. Often can be done in the same test as retrieval test. If we have this test, we can just check for other scenarios if id matches what we expect, and we DROurselfs :P

File: core/tests/test_views.py [hunk 6]
```diff
@@ -0,0 +1,659 @@
...
+@pytest.mark.django_db
+class TestGenre:
+    url_list = reverse("genre-list")
+
+    def test_genre_post(self, client):
+        genre = GenreFactory.build()
+        serialized = GenreSerializer(genre).data
+
+        response = client.post(
+            self.url_list, data=serialized, content_type="application/json"
```
Comment (maclewan) [inline core/tests/test_views.py]: if I recall correctly from top of my head, `application/json` is redundant, and client will default to it

File: core/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,659 @@
...
+    def test_genre_get_list(self, client):
+        genre1 = GenreFactory.create()
+        genre2 = GenreFactory.create()
+
+        response = client.get(self.url_list)
+        response_data = response.json()
+
+        assert response.status_code == 200
+        assert len(response_data) == 2
+        assert {"id", "name", "is_for_adults"} <= set(response_data[0].keys())
```
Comment (maclewan) [inline core/tests/test_views.py]: same as above, if we have one full serialization test, we can skip that kind of checks. Also this helps for easy test maintenance and single responsibility

File: core/tests/test_views.py [hunk 8]
```diff
@@ -0,0 +1,659 @@
...
+        genre = GenreFactory.create()
+        url = reverse("genre-detail", args=[genre.id])
+
+        response = client.delete(url)
+
+        assert response.status_code == 204
+        assert not Genre.objects.filter(id=genre.id).exists()
+
+
+@pytest.mark.django_db
```
Comment (maclewan) [inline core/tests/test_views.py]: you can do single pytest mark on top for django db, figure out how to do that

Comment (maclewan) [reply core/tests/test_views.py]: I am not saying it is the best practice, but most common

File: core/tests/test_views.py [hunk 9]
```diff
@@ -0,0 +1,659 @@
...
+
+        assert response.status_code == 200
+        assert response_data["name"] == actor.name
+
+    def test_actor_put(self, client):
+        actor = ActorFactory.create()
+        url = reverse("actor-detail", args=[actor.id])
+        new_data = {
+            "name": "Updated Actor",
+            "age": actor.age + 1,
```
Comment (maclewan) [inline core/tests/test_views.py]: maybe nitpicking maybe not. If we can avoid any state if test I would do that. If we can put fixed age into factory I would do that, so test always tests the same. Also this test might fail with validation error, if for some reason factory picks highest legal age value, and test tries updating it to +1

File: core/tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,659 @@
...
+            "nationality": actor.nationality,
+        }
+
+        response = client.put(url, data=new_data, content_type="application/json")
+        response_data = response.json()
+
+        assert response.status_code == 200
+        assert response_data["name"] == new_data["name"]
+        assert response_data["age"] == new_data["age"]
+        assert response_data["nationality"] == new_data["nationality"]
```
Comment (maclewan) [inline core/tests/test_views.py]: in case like that, to simplify test, you can just assert whole response data

File: core/tests/test_views.py [hunk 11]
```diff
@@ -0,0 +1,659 @@
...
+
+    def test_actor_delete(self, client):
+        actor = ActorFactory.create()
+        url = reverse("actor-detail", args=[actor.id])
+
+        response = client.delete(url)
+
+        assert response.status_code == 204
+        assert not Actor.objects.filter(id=actor.id).exists()
```
Comment (maclewan) [inline core/tests/test_views.py]: This will work, but if you know that value is False I would suggest asserting as follows:
` assert Actor.objects.filter(id=actor.id).exists() is False`. And do this everywhere as a habit, to avoid asserting things like empty list or dict, instead of False or None. Same for none, if you know something might be none, check `is None`

Comment (maclewan) [reply core/tests/test_views.py]: this might save you from some weird bugs

File: core/tests/test_views.py
Comment (maclewan) [inline core/tests/test_views.py:1]: Overal suggestions:
1. There are 2 common ways of placing tests in repo.
*   each django app has its `test` folder in it - I personally dont like that, it is messy to maintain, and hard to separate tests code
* there is `tests` folder alongside other apps folders, and inside there is dedicated folder for each app
pick whichever you like, but be sure you understand both approaches and can explain up and downsides of each

2.  To much tests in one file. Separate tests into at least model specific files, for some crazy tests and huge models, I have seen/done even file per action - `test_user_listing`, `test_user_updating`, and so on.

3. I would suggest one test for serialization. I tend to put such at the top, but it is you to you. This way, other tests (assuming they use the same serializer) can just check logic, not the serialization details

Comment (maclewan) [reply core/tests/test_views.py:1]: 4. I dont really like class based tests, because it can get tricky to understand setup and cleanup. with function based test, each test is it's own universe, and fixtures can be used, to implement reusable details

File: core/tests/test_views.py (hunk 10 above)
Comment (adlbrtbsr) [reply core/tests/test_views.py]: Are you sure? Requests also return objects ID which is not included in the new data

File: core/tests/test_views.py (hunk 6 above)
Comment (adlbrtbsr) [reply core/tests/test_views.py]: Not always, just checked
